---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```
# glmGamPoi

<!-- badges: start -->
<!-- badges: end -->

> Fit Small and Large Scale Gamma-Poisson Generalized Linear Models.

_Gamma-Poisson is just another name for the Negative Binomial model. It however better emphasizes the overdispersed count nature of this model._

`glmGamPoi` provides an easy to use interface to fit Gamma-Poisson models efficiently. It is designed to handle large scale datasets with thousands of rows and columns as they are typically encountered in modern high-throughput biology. It can automatically determine the appropriate size factors to normalize for different sizes across samples and infer the overdispersion for each row in the input matrix. The package can handle in-memory datasets, but also on-disk matrices with the `DelayedArray` package.


## Installation

You can install the development version of `glmGamPoi` from [Github](https://github.com/const-ae/glmGamPoi):

``` r
# install.packages("devtools")
devtools::install_github("const-ae/glmGamPoi")
```

Please make sure that you are using at least version `3.6` of R and `3.10` of BioConductor.

## Example

To fit a single Gamma-Poisson GLM do:

```{r}
# overdispersion = 1/size
counts <- rnbinom(n = 10, mu = 5, size = 1/0.7)
# size_factors = FALSE, because only a single GLM is fitted
glmGamPoi::glm_gp(counts, design = ~ 1, size_factors = FALSE)
```

The `glm_gp()` function returns a list with the results of the fit. Most importantly, it contains the estimates for the coefficients β and the overdispersion.

Fitting repeated Gamma-Poisson GLMs for each gene of a single cell dataset is just as easy:

```{r}
pbmcs <- TENxPBMCData::TENxPBMCData("pbmc4k")
pbmcs
# Fit the Gamma-Poisson Model. Force calculation to be in memory
fit <- glmGamPoi::glm_gp(pbmcs, on_disk = FALSE)
# First fitted values. -Inf indicates rows that contained only zeros.
head(fit$Beta_est)

plot(fit$Mu[,1], fit$overdispersions + 1e-6, 
     log ="xy", pch = 16, cex = 0.3,
     main = "Trend of Overdispersion against Mean")
```



## Benchmark

For demonstration purposes, I create a sample benchmark with 300 non-empty genes from the `pbmc4k` dataset:

```{r}
non_empty_rows <- which(DelayedMatrixStats::rowSums2(assay(pbmcs)) > 0)
pbmcs_subset <- as.matrix(SummarizedExperiment::assay(pbmcs)[sample(non_empty_rows, 300), ])
model_matrix <- matrix(1, nrow = ncol(pbmcs_subset))
dim(pbmcs_subset)
```


I compare my method running in-memory and on-disk with `DESeq2` and `edgeR`. Both are classical methods for analyzing RNA-Seq datasets and have been around for almost 10 years. (Note that both can do a lot more than just fitting the Gamma-Poisson model, so this benchmark only serves to give a general impression of the performance.)

```{r}
bench::mark(
  glmGamPoi_in_memory = {
    glmGamPoi::glm_gp(pbmcs_subset, design = model_matrix, on_disk = FALSE)
  }, glmGamPoi_on_disk = {
    glmGamPoi::glm_gp(pbmcs_subset, design = model_matrix, on_disk = TRUE)
  }, DESeq2 = suppressMessages({
    dds <- DESeq2::DESeqDataSetFromMatrix(pbmcs_subset, 
                        colData = data.frame(name = seq_len(4340)),
                        design = ~ 1)
    dds <- DESeq2::estimateSizeFactors(dds, "poscounts")
    dds <- DESeq2::estimateDispersions(dds, quiet = TRUE)
    dds <- DESeq2::nbinomWaldTest(dds, minmu = 1e-6)
  }), edgeR = {
    edgeR_data <- edgeR::DGEList(pbmcs_subset)
    edgeR_data <- edgeR::calcNormFactors(edgeR_data)
    edgeR_data <- edgeR::estimateDisp(edgeR_data, model_matrix)
    fit <- edgeR::glmFit(edgeR_data, design = model_matrix)
  }, check = FALSE
)
```

Fitting the full `pbmc4k` dataset on a modern MacBook Pro in-memory takes ~1 minute and on-disk ~4 minutes. Fitting the `pbmc68k` (17x the size) takes ~73 minutes (17x the time) on-disk. Fitting that dataset in-memory is not possible because it is just too big: the maximum in-memory matrix size is `2^31-1 ≈ 2.1e9` is elements, the `pbmc68k` dataset however has nearly 100 million more elements.


